### PR TITLE
Add a flag to specify distribution strategies.

### PR DESCRIPTION
**Pull request by yuefengz**

### CODE REVIEW

File: official/utils/misc/distribution_utils.py [hunk 1]
```diff
@@ -23,42 +23,47 @@
...
+        raise ValueError("`OneDeviceStrategy` can not be used for more than "
+                         "one device.")
       return tf.contrib.distribute.OneDeviceStrategy("device:GPU:0")
-  elif turn_off_distribution_strategy:
-    raise ValueError("When {} GPUs are specified, "
-                     "turn_off_distribution_strategy flag cannot be set to"
-                     "True.".format(num_gpus))
-  else:  # num_gpus > 1 and not turn_off_distribution_strategy
+
+  if distribute_strategy in ("mirrored", "default"):
```
Comment (guptapriya) [inline official/utils/misc/distribution_utils.py]: throw error here if num_gpus <=0 ?

File: official/utils/misc/distribution_utils.py [hunk 2]
```diff
@@ -68,6 +73,17 @@ def get_distribution_strategy(num_gpus,
...
+  if distribute_strategy == "collective":
+    return tf.contrib.distribute.CollectiveAllReduceStrategy(
+        num_gpus_per_worker=num_gpus)
+
+  if distribute_strategy == "parameter_server":
+    return tf.contrib.distribute.ParameterServerStrategy(
+        num_gpus_per_worker=num_gpus)
+
+  raise ValueError(
+      "Uncognized Distribution Strategy: %r" % distribute_strategy)
```
Comment (guptapriya) [inline official/utils/misc/distribution_utils.py]: "Unrecognized"

File: official/utils/misc/distribution_utils.py [hunk 3]
```diff
@@ -68,6 +73,17 @@ def get_distribution_strategy(num_gpus,
     else:
       return tf.distribute.MirroredStrategy(devices=devices)
 
+  if distribute_strategy == "collective":
+    return tf.contrib.distribute.CollectiveAllReduceStrategy(
```
Comment (guptapriya) [inline official/utils/misc/distribution_utils.py]: oh no :(
need these to be in tf core soooon...

File: official/utils/misc/distribution_utils.py (hunk 2 above)
Comment (yuefengz) [reply official/utils/misc/distribution_utils.py]: Done.

File: official/utils/misc/distribution_utils.py (hunk 1 above)
Comment (yuefengz) [reply official/utils/misc/distribution_utils.py]: Done. Added to the beginning. We can remove/relax it when we add TPUStrategy.

File: official/utils/misc/distribution_utils.py (hunk 3 above)
Comment (yuefengz) [reply official/utils/misc/distribution_utils.py]: Will update it shortly.

File: official/utils/misc/distribution_utils.py [hunk 4]
```diff
@@ -23,43 +23,51 @@
...
-                     'True.'.format(num_gpus))
-  else:  # num_gpus > 1 and not turn_off_distribution_strategy
-    devices = ['device:GPU:%d' % i for i in range(num_gpus)]
+      if num_gpus > 1:
+        raise ValueError("`OneDeviceStrategy` can not be used for more than "
+                         "one device.")
+      return tf.contrib.distribute.OneDeviceStrategy("device:GPU:0")
+
+  if distribution_strategy in ("mirrored", "default"):
+    devices = ["device:GPU:%d" % i for i in range(num_gpus)]
```
Comment (guptapriya) [inline official/utils/misc/distribution_utils.py]: the check i was asking for is actually this case: distribution_strategy=mirrored, and num_gpus=0. 
We will pass empty list to MirroredStrategy. what happens in that case? (not a big deal, just checking)

**Comment by guptapriya** [comment]
Feel free to merge this when you've coordinate with Toby and Shining.

**Comment by tfboyd** [comment]
@yuefengz   I hope my ask above is not too hard.  We want to get into a habit of updating the tests before they break.

File: official/utils/misc/distribution_utils.py (hunk 4 above)
Comment (yuefengz) [reply official/utils/misc/distribution_utils.py]: I thought MS would use cpu but turns out not. So updated the code to cover gpus=0 case.

**Comment by yuefengz** [comment]
@tfboyd Done. Also did a grep which couldn't find any occurrence of `turn_off_distribution_strategy` any more.

**Comment by tfboyd** [comment]
LGTM.  best of luck with merge conflicts.  Thank you for the extra effort.

**Comment by yuefengz** [comment]
Thanks for your review! I've tested it out and resolved the conflicts.

**Pull request merged**